File: src/CodexQuotaRail.AppServer/Resilience/BackoffSchedule.cs
namespace CodexQuotaRail.AppServer.Resilience;

public sealed class BackoffSchedule
{
    private static readonly TimeSpan[] Delays =
    [
        TimeSpan.FromSeconds(2),
        TimeSpan.FromSeconds(5),
        TimeSpan.FromSeconds(15),
        TimeSpan.FromSeconds(30),
        TimeSpan.FromSeconds(60),
    ];

    private int _index;

    public TimeSpan NextDelay()
    {
        var current = Math.Min(_index, Delays.Length - 1);
        if (_index < Delays.Length - 1)
        {
            _index++;
        }

        return Delays[current];
    }

    public void Reset() => _index = 0;
}

File: macos/Packages/CodexQuotaKit/Sources/CodexQuotaKit/BackoffSchedule.swift
public struct BackoffSchedule: Sendable {
    private static let delays: [Duration] = [
        .seconds(2),
        .seconds(5),
        .seconds(15),
        .seconds(30),
        .seconds(60),
    ]

    private var index = 0

    public init() {}

    public mutating func nextDelay() -> Duration {
        let delay = Self.delays[min(index, Self.delays.count - 1)]
        if index < Self.delays.count - 1 {
            index += 1
        }
        return delay
    }

    public mutating func reset() {
        index = 0
    }
}
